tcpping/tcpping.go

```diff
@@ -0,0 +1,95 @@
+package tcpping
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	errors "github.com/pkg/errors"
+	tcp "github.com/tevino/tcp-shaker"
+)
+
+type Result struct {
+	Address string
+	Latency int
+	PingErr error
+}
+
+func (cr *Result) String() string {
+	return fmt.Sprintf(
+		"Result{address: %s, latency: %d, error: %s}\n",
+		cr.Address, cr.Latency, cr.PingErr)
+}
+
+func runCheck(c *tcp.Checker, address string, resChan chan Result, timeout time.Duration) {
+	// mesaure RTT
+	start := time.Now()
+	// TCP Ping
+	err := c.CheckAddr(address, timeout)
+	// measure RTT
+	elapsed := time.Since(start)
+	latency := int(elapsed.Milliseconds())
+
+	if err != nil { // don't confuse users with valid latency values on error
+		latency = -1
+		if err == tcp.ErrTimeout {
+			err = errors.Wrap(err, "tcp check timeout")
+		} else {
+			switch err.(type) {
+			case tcp.ErrConnect:
+				err = errors.Wrap(err, "unable to connect")
+			default:
+				err = errors.Wrap(err, "unknown error")
+			}
+		}
+	}
+
+	// return results
+	resChan <- Result{
+		Address: address,
+		Latency: latency,
+		PingErr: err,
+	}
+}
+
+func waitForResults(resChan chan Result, expected int, timeout time.Duration) ([]Result, error) {
+	var checks = []Result{}
+	for {
+		select {
+		case result := <-resChan:
+			checks = append(checks, result)
+			if len(checks) == expected {
+				return checks, nil
+			}
+		case <-time.After(timeout):
+			return checks, errors.New("results wait timeout")
+		}
+	}
+}
+
+func CheckHosts(addresses []string, timeout time.Duration) ([]Result, error) {
+	c := tcp.NewChecker()
+
+	// stop the checking loop when function exists
+	ctx, stopChecker := context.WithCancel(context.Background())
+	defer stopChecker()
+
+	// loop that queries Epoll and pipes events
+	go func() {
+		// TODO handle error from this?
+		c.CheckingLoop(ctx)
+	}()
+
+	// channel for returning results from concurrent checks
+	var resChan = make(chan Result)
+
+	for i := 0; i < len(addresses); i++ {
+		go runCheck(c, addresses[i], resChan, timeout)
+	}
+
+	// add a bit extra time to let TCP connections time out
+	waitTimeout := timeout + (100 * time.Millisecond)
+
+	// wait for the results for all addresses
+	return waitForResults(resChan, len(addresses), waitTimeout)
```

Comment:
```suggestion
	return waitForResults(resChan)
```

No need timeout in `waitForResult` because `runCheck` goroutines have timeout and will exit before anyway. Also, they will run in parallel so the whole set should not run longer than `timeout`.

Reply:
done